From d46bb8e74dcf992c4f7eb840424f2339039e9fbb Mon Sep 17 00:00:00 2001
From: Mark Nudelman <markn@greenwoodsoftware.com>
Date: Tue, 29 Sep 2026 09:27:04 -0700
Subject: [PATCH] Fix use-after-free in expand_evars.

expand_evars finds instances of ${VAR} in a buffer and expands them to
the value of the env var by calling lgetenv("VAR") and appending the
result to an xbuf. But the value returned from lgetenv could be a
substring within the xbuf that we are working on. When appending to it,
we may free and reallocate the contents, invalidating the lgetenv
result. Save a copy of the lgetenv result separate from the xbuf.

This has always been a bug, but previously didn't happen because of
the way that expand_evars was invoked from the decode.c code. Change
5b0d521fe55aad4e4ce5d44a2b2bd0e4129e68b5 changes the way expand_evars
is called and makes the bug more likely to manifest.

Related to #831.
---
 evar.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/evar.c b/evar.c
index e723fa76..a6a29f70 100644
--- a/evar.c
+++ b/evar.c
@@ -171,6 +171,7 @@ public void expand_evars(mutable char *buf, size_t len, struct xbuffer *xbuf)
 		if (i+1 < len && buf[i] == '$' && buf[i+1] == '{')
 		{
 			constant char *evar;
+			char *evar2;
 			char term;
 			size_t e;
 			i += 2; /* skip "${" */
@@ -183,7 +184,11 @@ public void expand_evars(mutable char *buf, size_t len, struct xbuffer *xbuf)
 			buf[e++] = '\0';
 			evar = lgetenv_ext(&buf[i], xbuf->data, xbuf->end);
 			if (evar == NULL) evar = "";
-			i = add_evar(xbuf, buf, len, e, evar, term);
+			/* Need to save a copy of evar because it might be part of xbuf
+			 * and could get freed when we add to xbuf in add_evar. */
+			evar2 = save(evar);
+			i = add_evar(xbuf, buf, len, e, evar2, term);
+			free(evar2);
 		} else
 		{
 			xbuf_add_char(xbuf, buf[i++]);
